The ISA-L EC code has been written using RVV vector instructions and the minimum multiplication table, resulting in a performance improvement of over 10 times compared to the existing implementation.

lvshuo2016 opened the pull request:
Hi Pablo
     I rewrote the ISA-L EC code using the vector instructions on risc-v platform and complete the all function and performance test.
1. It uses vector instructions and the minimum multiplication table to perform table lookup operations. By leveraging vector instructions and the minimum multiplication table, multiple lookup or computation operations are executed within a single instruction cycle and the number of table lookups and calculations are reduced , so the cache miss rate significantly is reduced.  In existing implementation, to obtain log₂a and log₂b, two table lookups are performed, followed by one addition, and then one more table lookup for the exponential table. Including the final summation operation, generating a checksum block requires 3 * n table lookups and 2 * n calculations. So it's performance is low.
2. The performance has improved by more than 10 times compared to the the existing rvv implementation,  its performance is improved by more than 20 times compared to the the existing C implementation
3. The code is also more readable. 

current rvv:      
        erasure_code_encode_warm: runtime =    3065696 usecs, bandwidth 108 MB in 3.0657 sec = 650.61 MB/s         
       erasure_code_decode_warm: runtime =    3001213 usecs, bandwidth 136 MB in 3.0012 sec = 829.70 MB/s
        gf_vect_mul_warm: runtime =    3062512 usecs, bandwidth 6445 MB in 3.0625 sec = 2104.76 MB/s
        gf_vect_dot_prod_warm: runtime =    3002001 usecs, bandwidth 4027 MB in 3.0020 sec = 1341.56 MB/s
        ec_encode_data_update_warm: runtime =    3000191 usecs, bandwidth 2374 MB in 3.0002 sec = 791.39 MB/s
        ec_encode_data_update_single_src_warm: runtime =    3062524 usecs, bandwidth 6348 MB in 3.0625 sec = 2072.83 MB/s
        ec_encode_data_update_single_src_simple_warm: runtime =    3061624 usecs, bandwidth 8664 MB in 3.0616 sec = 2830.10 MB/s
        ec_encode_data_update_decode_warm: runtime =    3062507 usecs, bandwidth 1442 MB in 3.0625 sec = 471.06 MB/s

existing rvv:
        erasure_code_encode_warm: runtime =    3065696 usecs, bandwidth 108 MB in 3.0657 sec = 35.61 MB/s         
        erasure_code_decode_warm: runtime =    3001213 usecs, bandwidth 136 MB in 3.0012 sec = 45.70 MB/s
        gf_vect_mul_warm: runtime =    3062512 usecs, bandwidth 6445 MB in 3.0625 sec = 622.18 MB/s
        gf_vect_dot_prod_warm: runtime =    3002001 usecs, bandwidth 4027 MB in 3.0020 sec = 159.56 MB/s
        ec_encode_data_update_warm: runtime =    3000191 usecs, bandwidth 2374 MB in 3.0002 sec = 31.39 MB/s
        ec_encode_data_update_single_src_warm: runtime =    3062524 usecs, bandwidth 6348 MB in 3.0625 sec = 110.02 MB/s
        ec_encode_data_update_single_src_simple_warm: runtime =    3061624 usecs, bandwidth 8664 MB in 3.0616 sec = 110.10 MB/s
        ec_encode_data_update_decode_warm: runtime =    3062507 usecs, bandwidth 1442 MB in 3.0625 sec = 19.06 MB/s

c: 
        erasure_code_encode_warm: runtime =    3002512 usecs, bandwidth 52 MB in 3.0025 sec = 15.65 MB/s         erasure_code_decode_warm: runtime =    3065235 usecs, bandwidth 57 MB in 3.0652 sec = 18.19 MB/s
        gf_vect_mul_warm: runtime =    3063951 usecs, bandwidth 307 MB in 3.0640 sec = 101.05 MB/s
        gf_vect_dot_prod_warm: runtime =    3010048 usecs, bandwidth 180 MB in 3.0100 sec = 60.02 MB/s
        ec_encode_data_update_warm: runtime =    3000191 usecs, bandwidth 2374 MB in 3.0002 sec = 30.98 MB/s
        ec_encode_data_update_single_src_warm: runtime =    3062524 usecs, bandwidth 6348 MB in 3.0625 sec = 109.97 MB/s
        ec_encode_data_update_single_src_simple_warm: runtime =    3061624 usecs, bandwidth 8664 MB in 3.0616 sec = 110.43 MB/s
        ec_encode_data_update_decode_warm: runtime =    3062507 usecs, bandwidth 1442 MB in 3.0625 sec = 19.51 MB/s

Looking forward to your reply.

pablodelara commented:
Thanks. There are some formatting issues, can you fix those? There is a tool "./tools/format.sh" that you can run for that.
Also, can you use your full name in the "Signed-off-by" line? Thanks!

lvshuo2016 commented:
Hi Pablo 
    I updated the compilation issue because my terminal environment uses the export CFLAGS="-march=rv64gcv"
and export ASFLAGS="-march=rv64gcv" commands, which are not available in your environment. Please help me continue.

pablodelara commented on erasure_code/riscv64/Makefile.am:
```diff
@@ -1,5 +1,5 @@
-########################################################################
-#  Copyright (c) 2025 Institute of Software Chinese Academy of Sciences (ISCAS).
+##################################################################
+#  Copyright (c) 2025 sanechips Technologies Co., Ltd.
```
Sorry, you cannot remove the Copyright from original authors

pablodelara commented on erasure_code/riscv64/ec_encode_data_rvv.S:
```diff
@@ -1,154 +0,0 @@
-.option         arch, +v
-.global         ec_encode_data_rvv
-.type           ec_encode_data_rvv, %function
-ec_encode_data_rvv:
```
I have a fundamental question. Your new code is overwriting existing implementation of EC encode. Is the new implementation only available for a specific RISC-V architecture (newer than the previous one)?

lvshuo2016 replied:
> specific RISC-V architecture



> I have a fundamental question. Your new code is overwriting existing implementation of EC encode. Is the new implementation only available for a specific RISC-V architecture (newer than the previous one)?

It is universal for all implementations supporting RVV.  If RV does not support vector instructions, the code could not run

lvshuo2016 replied:
> It is universal for all implementations supporting RVV. If RV does not support vector instructions, the code could not run

This code has been tested on version 2.30 for a long time. Users need to test on this specific version. Recently, the patch was directly applied to 2.31. On RV, due to memory address alignment problems, some tests case cannot pass. I will check whether all the patches have been ported over.

lvshuo2016 commented on configure.ac:
```diff
@@ -69,6 +69,8 @@ case "${CPU}" in
 			AM_CONDITIONAL([HAVE_RVV], [true]) rvv=yes],
 			[AM_CONDITIONAL([HAVE_RVV], [false]) rvv=no]
 		)
+		CFLAGS+=" -march=rv64gcv"
```
If RV could not support vector instruction, it could not run. I will add a macro definition. What do you think?

pablodelara replied:
These tests are run in an actual RISC-V machine provided by ISCAS. @sunyuechi provided that machine. It would be good if they can add information of that environment.

lvshuo2016 replied:
It seems I know what the problem is. The latest compilation has the O2 option enabled, but my previous test did not. O2 causes buffs[0] to be NULL. in erasure_code/erasure_code_test.c  if (0 != memcmp(temp_buffs[k + i], buffs[src_err_list[i]], TEST_LEN)), Maybe the assembly code is optimized when 02 is selected.

lvshuo2016 replied:
I asked the compiler team, and they said that there were some issue with GCC12 under RV when use C and assembly code, and suggested to use Gcc 14 version.

pablodelara commented:
Can you share the ISA info? Looks like the Ci machine is also Spacemit X60:

model name : Spacemit(R) X60
isa : rv64imafdcv_zicbom_zicboz_zicntr_zicond_zicsr_zifencei_zihintpause_zihpm_zfh_zfhmin_zca_zcd_zba_zbb_zbc_zbs_zkt_zve32f_zve32x_zve64d_zve64f_zve64x_zvfh_zvfhmin_zvkt_sscofpmf_sstc_svinval_svnapot_svpbmt



From: lvshuo2016 ***@***.***>
Sent: Friday, June 6, 2025 2:39 PM
To: intel/isa-l ***@***.***>
Cc: De Lara Guarch, Pablo ***@***.***>; Mention ***@***.***>
Subject: Re: [intel/isa-l] The ISA-L EC code has been written using RVV vector instructions and the minimum multiplication table, resulting in a performance improvement of over 10 times compared to the existing implementation. (PR #331)


@lvshuo2016 commented on this pull request.

________________________________

In configure.ac<https://github.com/intel/isa-l/pull/331#discussion_r2132219980>:

> @@ -69,6 +69,8 @@ case "${CPU}" in

                       AM_CONDITIONAL([HAVE_RVV], [true]) rvv=yes],

                       [AM_CONDITIONAL([HAVE_RVV], [false]) rvv=no]

               )

+              CFLAGS+=" -march=rv64gcv"

In my ***@***.*** isa-l]$ lscpu
Architecture: riscv64
Byte Order: Little Endian
CPU(s): 8
On-line CPU(s) list: 0-7
Model name: Spacemit(R) X60
Thread(s) per core: 1
Core(s) per socket: 8
Socket(s): 1
Frequency boost: disabled
CPU(s) scaling MHz: 100%
CPU max MHz: 1600.0000
CPU min MHz: 614.4000
Caches (sum of all):
L1d: 256 KiB (8 instances)
L1i: 256 KiB (8 instances)
L2: 1 MiB (2 instances)
all test passed.
In our EMU test environment,it is also passed.

—
Reply to this email directly, view it on GitHub<https://github.com/intel/isa-l/pull/331#discussion_r2132219980>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AFPDEPP7KMG4BHXIARDRP3D3CGKYFAVCNFSM6AAAAAB6PAWY4SVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDSMBUHE3DANBYGY>.
You are receiving this because you were mentioned.Message ID: ***@***.******@***.***>>

lvshuo2016 commented:
[root@openeuler-riscv64 isa-l-ec-vector-branch]# gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/riscv64-openEuler-linux/12/lto-wrapper
Target: riscv64-openEuler-linux
Configured with: ../configure --disable-bootstrap --disable-libgcj --without-cloog --enable-languages=c,c++,fortran,objc,obj-c++,lto --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://gitee.com/src-openeuler/gcc/issues --enable-shared --enable-threads=posix --enable-checking=release --disable-multilib --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-gcc-major-version-only --enable-libstdcxx-backtrace --with-linker-hash-style=gnu --enable-plugin --enable-initfini-array --without-isl --with-arch=rv64gc --with-abi=lp64d --disable-libquadmath --build=riscv64-openEuler-linux
Thread model: posix
Supported LTO compression algorithms: zlib
gcc version 12.3.1 (openEuler 12.3.1-62.oe2403sp1) (GCC)
[root@openeuler-riscv64 isa-l-ec-vector-branch]#                   It is related to the gcc compiler optimization under risc-v. I will feedback to our company's compiler team to track it separately.

sunyuechi commented on include/riscv64_multibinary.h:
This part was removed in https://github.com/intel/isa-l/commit/7e01b2c81222df170a8462594da25ab64acf28d8. Should it be removed as well?

lvshuo2016 replied:
Thank you very much for your reminder.  I found that there was no similar warning message during compilation. Could I not remove it?

lvshuo2016 replied:
This part has been removed.

pablodelara commented on erasure_code/erasure_code_update_test.c:
```diff
@@ -59,6 +59,11 @@
 
+#if HAVE_RVV
+#define EC_ALIGNED_ADDR
+#define ALIGN_SIZE 32
```
Can you move this ALIGN_SIZE to line 38? So:

`
#if HAVE_RVV
#define ALIGN_SIZE 32
#else
#define ALIGN_SIZE 16
#endif
`

pablodelara commented:
> @pablodelara I ran it and passed the test, which had a good performance improvement, but I haven't looked at the specific implementation carefully, I just saw some possible format/build problems.

Thanks for checking, @sunyuechi. Please, report these problems, so they can be fixed before this is integrated.

sunyuechi commented:
@lvshuo2016  I think we should ensure that the code can be correctly compiled and run regardless of whether the machine and compiler support the V extension. Currently, there is runtime detection, but some HAVE_RVV checks have been removed, which leads to build errors if the compiler does not support the V extension.

https://github.com/intel/isa-l/blob/8054d41db5efa63d49e942880c4b53be7c4838c8/configure.ac#L65

You can temporarily replace this with an instruction that fails to compile in order to test building with a compiler that doesn't support the V extension.

lvshuo2016 commented:
> One comment from me in the code. Also, can you reword the commit title and remove your second signed-off-by line? Also, check release notes and see if you can extend it At the moment, it says:
> 
> * Added new RVV ec_encode_data, gf_vect_dot_prod, gf_vect_mul implementations.
>   But this is adding other implementations too.

I will modify it.

lvshuo2016 commented:
> @lvshuo2016 I think we should ensure that the code can be correctly compiled and run regardless of whether the machine and compiler support the V extension. Currently, there is runtime detection, but some HAVE_RVV checks have been removed, which leads to build errors if the compiler does not support the V extension.
> 
> https://github.com/intel/isa-l/blob/8054d41db5efa63d49e942880c4b53be7c4838c8/configure.ac#L65
> 
> You can temporarily replace this with an instruction that fails to compile in order to test building with a compiler that doesn't support the V extension.

Thanks, I will add HAVE_RVV check.

pablodelara commented on erasure_code/erasure_code_update_test.c:
```diff
@@ -35,8 +35,13 @@
 #ifndef ALIGN_SIZE
+#if HAVE_RVV
+#define EC_ALIGNED_ADDR
+#define ALIGN_SIZE  32
```
All you need is to remove the extra whitespace to fix the format:

#define ALIGN_SIZE 32

lvshuo2016 replied:
Thanks. Already modified.

pablodelara commented on erasure_code/riscv64/Makefile.am:
```diff
@@ -28,8 +29,23 @@
-	erasure_code/riscv64/ec_encode_data_rvv.S
+		erasure_code/riscv64/ec_riscv64_dispatcher.c \
+		erasure_code/riscv64/ec_multibinary_riscv64.S
+if HAVE_RVV
```
Missing `#` here in endif below.

pablodelara replied:
This breaks most of the CI tests:
erasure_code/riscv64/Makefile.am:34: *** missing separator.  Stop.

Can you take a look and fix it?

lvshuo2016 replied:
Could CI tests be run locally on my machine?

lvshuo2016 replied:
Looks it need add a new line before if in Makefike.am.

lvshuo2016 replied:
I saw that the Makefile.am in the root directory also had a similar if, and it was fine if there is a new  line in front of if.

pablodelara replied:
Cannot you see the issue compiling with "make -f Makefile.unx" ?

pablodelara commented on erasure_code/riscv64/ec_riscv64_highlevel_func.c:
```diff
@@ -0,0 +1,216 @@
+                break;
+        }
+}
+/*
```
This code can be removed, right?

lvshuo2016 commented:
Hi pablo，all tests have passed.

lvshuo2016 commented on erasure_code/erasure_code_test.c:
```diff
@@ -50,6 +50,9 @@
 #define EFENCE_TEST_MIN_SIZE 16
 #define EFENCE_TEST_MAX_SIZE EFENCE_TEST_MIN_SIZE + 0x100
 
+#if HAVE_RVV
```
Thanks.  I add one line in configure.ac instead of replacing #if , what do you think?
AC_COMPILE_IFELSE(
    [AC_LANG_PROGRAM([], [
        __asm__ volatile(
            ".option arch, +v\n"
            "vsetivli zero, 0, e8, m1, ta, ma\n"
        );
    ])],
    [AC_DEFINE([HAVE_RVV], [1], [Enable RVV instructions])
     AM_CONDITIONAL([HAVE_RVV], [true]) rvv=yes],
    [AC_DEFINE([HAVE_RVV], [0], [Disable RVV instructions])  <<<< add a new line
     AM_CONDITIONAL([HAVE_RVV], [false]) rvv=no]
)

lvshuo2016 replied:
modified

pablodelara replied:
So basically, #if HAVE_RVV  is the same as #if HAVE_RVV == 1.

lvshuo2016 replied:
yes, if the value of HAVE_RVV could only be 0 or 1. they are the same.

lvshuo2016 commented on erasure_code/riscv64/gf_3vect_dot_prod_rvv.S:
```diff
@@ -0,0 +1,197 @@
+########################################################################
+
+
+# RISC-V V扩展实现 gf_3vect_dot_prod_sve
+# 函数原型：
+# void gf_3vect_dot_prod_sve(int len, int vlen, unsigned char *gftbls,
+#                            unsigned char **src, unsigned char **dest);
```
Modified.

lvshuo2016 commented on erasure_code/riscv64/Makefile.am:
```diff
@@ -1,5 +1,6 @@
-########################################################################
-#  Copyright (c) 2025 Institute of Software Chinese Academy of Sciences (ISCAS).
+###########################################################################
+#  Copyright (c) 2025 Institute of Software Chinese Academy of Sciences (ISCAS)
```
This file has been completely rewritten. I will replace the original file.

lvshuo2016 commented on erasure_code/riscv64/gf_4vect_mad_rvv.S:
```diff
@@ -0,0 +1,197 @@
+.option norvc
+
+.global gf_4vect_mad_rvv
+#ifndef __APPLE__
+.type gf_4vect_mad_rvv, %function
+#endif
```
Thanks. Removed.

lvshuo2016 commented on erasure_code/riscv64/gf_vect_dot_prod_rvv.S:
```diff
@@ -0,0 +1,139 @@
+    ld t1, 0(t6)              # Load pointer to current source vector
+
+    # Load src data into z_src
+#    vsetvli t5, a0, e8, m1    # Set vector length based on len
```
removed

lvshuo2016 commented on erasure_code/riscv64/gf_vect_mul_rvv.S:
```diff
@@ -0,0 +1,122 @@
+    vmv.v.i v_mask0f, 0x0F   /* v_mask0f = 0x0F0F...0F */
+
+    /* Load pre-calculated constants into v_gft1_lo and v_gft1_hi */
+   // vsetivli x0, 16, e8, m1  /* Set vector length to 16 bytes */
```
removed.

lvshuo2016 commented on erasure_code/riscv64/gf_vect_mul_rvv.S:
```diff
@@ -0,0 +1,122 @@
+
+.Lloopsve_vl:
+    /* Set vector length based on remaining bytes */
+    //sub t2, x_len, x_pos
```
removed

pablodelara commented:
@sunyuechi are you OK with the current RISC-V code, given that you were the first one that upstreamed a RVV implementation?

sunyuechi commented:
> @sunyuechi are you OK with the current RISC-V code, given that you were the first one that upstreamed a RVV implementation?

It looks like my review comments on configure.ac and erasure_code/riscv64/ec_riscv64_dispatcher.c haven't been addressed or replied to yet.

lvshuo2016 commented:
> we should ensure that the code can be correctly compiled and run regardless of whether the machine and compiler support the V extension. Currently, there is runtime detection, bu

I have 

> @lvshuo2016 I think we should ensure that the code can be correctly compiled and run regardless of whether the machine and compiler support the V extension. Currently, there is runtime detection, but some HAVE_RVV checks have been removed, which leads to build errors if the compiler does not support the V extension.
> 
> https://github.com/intel/isa-l/blob/8054d41db5efa63d49e942880c4b53be7c4838c8/configure.ac#L65
> 
> You can temporarily replace this with an instruction that fails to compile in order to test building with a compiler that doesn't support the V extension.

thanks for your suggestion.

Sorry, forgot to reply, this was changed and tested a long time ago

lvshuo2016 commented:
> > @sunyuechi are you OK with the current RISC-V code, given that you were the first one that upstreamed a RVV implementation?
> 
> It looks like my review comments on configure.ac and erasure_code/riscv64/ec_riscv64_dispatcher.c haven't been addressed or replied to yet.

Sorry, the reply was missed, all the modifications were tested long ago. The code can be correctly compiled and run regardless of whether the machine and compiler support the V extension, and the HAVE_RVV macro has been added.

lvshuo2016 commented:
> > > we should ensure that the code can be correctly compiled and run regardless of whether the machine and compiler support the V extension. Currently, there is runtime detection, bu
> > 
> > 
> > I have
> > > @lvshuo2016 I think we should ensure that the code can be correctly compiled and run regardless of whether the machine and compiler support the V extension. Currently, there is runtime detection, but some HAVE_RVV checks have been removed, which leads to build errors if the compiler does not support the V extension.
> > > https://github.com/intel/isa-l/blob/8054d41db5efa63d49e942880c4b53be7c4838c8/configure.ac#L65
> > > 
> > > You can temporarily replace this with an instruction that fails to compile in order to test building with a compiler that doesn't support the V extension.
> > 
> > 
> > thanks for your suggestion.
> > Sorry, Reply missing, this was changed and tested a long time ago

I changed vsetvli to unknown command to test it.

lvshuo2016 commented:
> One of the if and fi statements is indented with spaces, and the other with tabs

One of the if and fi statements is indented with spaces, and the other with tabs, Which file or all files have this problem? I will fix it.


> @lvshuo2016
> 
> I was referring to the review on the GitHub files... These two parts haven't changed:
> 
> > One of the if and fi statements is indented with spaces, and the other with tabs, which causes them to appear misaligned in git diff
> 
> > This file should have been modified from ec_multibinary_riscv64_dispatcher.c. Currently, both ec_multibinary_riscv64_dispatcher.c and ec_riscv64_dispatcher.c are present in the directory, but ec_multibinary_riscv64_dispatcher.c is not included in the Makefile.

One of the if and fi statements is indented with spaces,  is the problem with all files or just one file? I will fix it.

At that time, ec_riscv64_dispatcher.c was completely rewritten and the file name was consistent with the previous architecture. I didn't know that there was a part of the code before. I will add origal Copyright to the front of the ec_riscv64_dispatcher.c，

lvshuo2016 commented:
Hi Pablo, Could the code be merged?

sunyuechi commented:
@lvshuo2016 

A few days ago, there was a review comment on `erasure_code/riscv64/gf_3vect_mad_rvv.S`. Could you please reply to it? I'm not sure if GitHub notified you about these review comments.

lvshuo2016 commented:
> A few days ago, there was a review comm

I didn't receive these review comments. Can you tell me the link? I haven't seen it, including the one you mentioned last time.

sunyuechi commented:
@lvshuo2016 It seems there's no direct link here for jumping to the location. You can scroll up on the PR page or check in the commit file to find it. I've attached a screenshot here.

![1](https://github.com/user-attachments/assets/2adc3332-c09d-4372-be0d-4b9b4cc8c37d)

lvshuo2016 commented:
> From my perspective, this commit includes several cases where the original files were directly deleted and replaced with new ones, which could have been handled more clearly.
> 
> For example, in `erasure_code/riscv64/Makefile.am`, although it was rewritten from scratch locally, the usual approach should be to modify the original file and resolve any conflicts during the commit, rather than deleting and replacing the file.
> 
> This has resulted in the original copyright being completely removed, and even the tabs have been changed.
> 
> I noticed the earlier discussion regarding `erasure_code/riscv64/Makefile.am`:
> 
> > Sorry, you cannot remove the Copyright from original authors
> > I have kept the original authors. But the file is fully rewritten. It's totally different from before, I added a separate line to explain.
> 
> However, the latest commit still does not include the original copyright.
> 
> If changes to tabs or file renaming are necessary, it might be better to add a separate commit specifically for the refactoring.
> 
> For instance, files like `ec_multibinary_riscv64.S` and `ec_riscv64_dispatcher.c` were renamed or deleted directly. Mixing these changes into a performance optimization commit makes things confusing.

Let me explain the background. In fact, it is not a performance optimization. The entire implementation architecture and method are completely different. This has been coded since last year, and all coding and testing have been completed in April this year, so there is  no modifications based on your code.  I can add the copyright of Makefile.am, and the others will remain unchanged.

lvshuo2016 commented:
> > From my perspective, this commit includes several cases where the original files were directly deleted and replaced with new ones, which could have been handled more clearly.
> > For example, in `erasure_code/riscv64/Makefile.am`, although it was rewritten from scratch locally, the usual approach should be to modify the original file and resolve any conflicts during the commit, rather than deleting and replacing the file.
> > This has resulted in the original copyright being completely removed, and even the tabs have been changed.
> > I noticed the earlier discussion regarding `erasure_code/riscv64/Makefile.am`:
> > > Sorry, you cannot remove the Copyright from original authors
> > > I have kept the original authors. But the file is fully rewritten. It's totally different from before, I added a separate line to explain.
> > 
> > 
> > However, the latest commit still does not include the original copyright.
> > If changes to tabs or file renaming are necessary, it might be better to add a separate commit specifically for the refactoring.
> > For instance, files like `ec_multibinary_riscv64.S` and `ec_riscv64_dispatcher.c` were renamed or deleted directly. Mixing these changes into a performance optimization commit makes things confusing.
> 
> Let me explain the background. In fact, it is not a performance optimization. The entire implementation architecture and method are completely different. This has been coded since last year, and all coding and testing have been completed in April this year, so there is no modifications based on previous code. I can add the copyright of Makefile.am, and the others will remain unchanged. At the beginning of the year, the department has been applying for the company's open source process approval, or else it would have been submitted long ago.

or add a copyright in file ec_multibinary_riscv64.S?

lvshuo2016 commented:
@pablodelara @sunyuechi @tkanteck   All questions have been replied, please check if there are any missing. If not, the next step is to merge.

pablodelara commented:
@lvshuo2016 why are you closing this PR?

lvshuo2016 commented:
> @lvshuo2016 why are you closing this PR?

I originally wanted to mark the issue status as resolved. If it is not correct, I will rewrite it to open

pablodelara commented:
> > @lvshuo2016 why are you closing this PR?
> 
> I meant to mark the issue status as resolved. If that's wrong, I'll reopen it.

This is a PR, not an issue, so closing the PR would indicate that you don't want it to be merged anymore.
@sunyuechi are you happy with the PR now?

sunyuechi commented:
@pablodelara 

Hi,

I think this should be decided by the project maintainer. As in the discussion above, 
for code with the same functionality but different implementations, the submission deletes and replaces the file with a new one.

I'm not sure whether such cases should still retain the original copyright.

For example, the latest commit revert the copyright in `erasure_code/riscv64/Makefile.am`, but not in `erasure_code/riscv64/gf_vect_mul_rvv.S`.

Also, I’ve only briefly looked at some of the maintenance-related issues and haven’t examined the implementation in detail yet.

If the above issues are resolved, I’ll review the implementation details as soon as possible.

@lvshuo2016 

There is a new comment in `gf_4vect_mad_rvv.S`

sunyuechi commented:
> Its implementation framework and ideas are completely different，so they are only the same file name for gf_vect_mul_rvv.S. It has been more than a month since I submitted it. I think understanding the implementation method and details will help to understand the copyright. If the methods are different, there is no point in worrying about copyright。
> 

Maybe I didn't express myself clearly.

> From my perspective, this commit includes several cases where the original files were directly deleted and replaced with new ones, which could have been handled more clearly.

What I meant is that the copyright is just one example — the real issue is that conflicts should be resolved and commits should be split appropriately.

For instance, changing tabs in `Makefile.am`, renaming `ec_multibinary_riscv64_dispatcher.c`, and whether the copyright should be removed — 
these caused confusion, which could have been avoided by splitting the commit to make things clearer.

lvshuo2016 commented:
> ![2](https://private-user-images.githubusercontent.com/89069008/460491358-9975b261-0764-451c-8761-97f682c3f9ff.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTEyNzQ3NzcsIm5iZiI6MTc1MTI3NDQ3NywicGF0aCI6Ii84OTA2OTAwOC80NjA0OTEzNTgtOTk3NWIyNjEtMDc2NC00NTFjLTg3NjEtOTdmNjgyYzNmOWZmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjMwVDA5MDc1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIwMGY3NTM4ZTk2MzhjMzAxMDAzOTY1ZGM1ODUyOWY0ZWEzNjliMGM2MzQ0MTc2MTVhM2QxZTJhNTJjNDgyYzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ditMxplsd3_PIfZYgiw9PnzA0j3gb6w1mS6mgsmCfIY)

I will modify it.

sunyuechi commented on erasure_code/riscv64/ec_multibinary_riscv64.S:
```diff
@@ -29,16 +57,10 @@
+mbin_interface gf_vect_mad
+mbin_interface ec_encode_data_update
+mbin_interface ec_init_tables
 
```
There is a blank line at the end of the file.

lvshuo2016 replied:
Thanks,modified.

lvshuo2016 commented on erasure_code/riscv64/gf_2vect_dot_prod_rvv.S:
```diff
@@ -0,0 +1,177 @@
+#define x_dest a4
+
+/* returns */
+//#define w_ret a0
```
Thanks.Modified.

sunyuechi commented on erasure_code/riscv64/gf_3vect_dot_prod_rvv.S:
```diff
@@ -0,0 +1,200 @@
+#                            unsigned char **src, unsigned char **dest);
+
+#if HAVE_RVV
+/* arguments */
```
The file has a different style compared to other `dot_prod` files (e.g., the order of .text and #define, and whether the comments refer to RVV or SVE). Please make them consistent.

lvshuo2016 replied:
Thanks.Modified.

sunyuechi commented on erasure_code/riscv64/gf_4vect_mad_rvv.S:
```diff
@@ -0,0 +1,194 @@
+ */
+/* arguments */
+#define x_len		a0
+#define x_vec		a1
```
The file uses different spacing after #define compared to other `vect_mad` files. Please make the style consistent.

sunyuechi commented on erasure_code/riscv64/gf_4vect_mad_rvv.S:
```diff
@@ -0,0 +1,194 @@
+
+	li	x_pos, 0
+
+.Lloopsve_vl:
```
Many of the SVE labels can be replaced with RVV or another more appropriate name.

lvshuo2016 replied:
Does the labels here say something like Lloopsve_vl?

sunyuechi commented on erasure_code/riscv64/gf_4vect_mad_rvv.S:
```diff
@@ -0,0 +1,194 @@
+.type gf_4vect_mad_rvv, %function
+
+/* gf_4vect_mad_rvv(int len, int vec, int vec_i, unsigned char *gftbls,
+			     unsigned char *src, unsigned char **dest);
```
The indentation here is inconsistent with other `mad_rvv` files. Please make them consistent.

sunyuechi replied:
This file still mixes tabs and spaces, whereas the other `mad_rvv` files use only spaces.

sunyuechi commented on erasure_code/riscv64/gf_3vect_dot_prod_rvv.S:
```diff
@@ -0,0 +1,207 @@
+
+# RISC-V Vector extension implementation gf_3vect_dot_prod_rvv
+# void gf_3vect_dot_prod_rvv(int len, int vlen, unsigned char *gftbls,
+#                            unsigned char **src, unsigned char **dest);
```
The `gf_3vect_dot_prod_rvv` here is now redundant with the following lines.

lvshuo2016 replied:
Thanks. Modified.

sunyuechi commented on erasure_code/riscv64/gf_4vect_mad_rvv.S:
```diff
@@ -0,0 +1,194 @@
+.align 2
+
+.global gf_4vect_mad_rvv
+.type gf_4vect_mad_rvv, %function
```
Other RVV files use `@function`, please keep it consistent

sunyuechi commented on erasure_code/riscv64/gf_3vect_dot_prod_rvv.S:
```diff
@@ -0,0 +1,207 @@
+    vle8.v v_src, (x_ptr)
+
+    /* Split 4-bit lo; 4-bit hi */
+    vand.vv v_src_lo, v_src, v_mask0f
```
`vmv.v.i + vand.vv` -> `vand.vi`, please check all `vand.vv`

sunyuechi commented on erasure_code/riscv64/gf_vect_mad_rvv.S:
```diff
@@ -0,0 +1,129 @@
+ */
+/* arguments */
+#define x_len a0
+#define x_vec a1
```
unused

lvshuo2016 replied:
Removed.

sunyuechi commented on erasure_code/riscv64/gf_3vect_dot_prod_rvv.S:
```diff
@@ -0,0 +1,201 @@
+    add x_dest2,x_dest2, x24
+    add x_dest3,x_dest3, x24
+
+    # 更新位置计数器
```
Remove all Chinese comments?

lvshuo2016 replied:
Modified.

sunyuechi commented on erasure_code/riscv64/gf_3vect_dot_prod_rvv.S:
```diff
@@ -0,0 +1,201 @@
+#  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+#  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+########################################################################
+# RISC-V Vector extension implementation gf_3vect_dot_prod_rvv
```
Other dot_prod_rvv files do not have this line; please keep it consistent

lvshuo2016 replied:
Removed.

sunyuechi commented on erasure_code/riscv64/gf_2vect_mad_rvv.S:
```diff
@@ -0,0 +1,153 @@
+
+    li x_pos, 0
+
+    /* vector length agnostic */
```
It seems the vector length condition is the same, but there are two places with this comment for some reason? Please make them consistent.

lvshuo2016 replied:
Modified.

sunyuechi commented on erasure_code/riscv64/gf_5vect_dot_prod_rvv.S:
```diff
@@ -0,0 +1,253 @@
+    j .Llooprvv_vl
+
+.return_pass:
+    /* Restore s0-s4 */
```
it restores s0–s10, but the comment says s0–s4.

lvshuo2016 replied:
Modified.

lvshuo2016 commented:
Now all is updated and all test passed.

lvshuo2016 commented on erasure_code/riscv64/gf_5vect_dot_prod_rvv.S:
```diff
@@ -0,0 +1,250 @@
+    sd s9, 56(sp)
+    sd s10, 64(sp)
+
+    li t0, 0x0F
```
Modified

lvshuo2016 commented on erasure_code/riscv64/gf_4vect_dot_prod_rvv.S:
```diff
@@ -0,0 +1,228 @@
+#define x_tbl2 t4
+#define x_tbl3 t5
+#define x_tbl4 t6
+##define x_dest1 t7
```
Removed.

lvshuo2016 commented on erasure_code/riscv64/gf_2vect_dot_prod_rvv.S:
```diff
@@ -0,0 +1,171 @@
+/* end of Loop 1 */
+
+.return_pass:
+    ld s8, 0(sp)
```
s8-s11 is equal to x24-x27. Initially the code is compiled is qemu, but the s8-s11 is not recognized by compiler. So initially x24-x27 is written. It is best not to make them consistent now. Once the compiler is completely stable, we can unify these registers s8-s11 to avoid repeated changes.. Once the compiler is completely stable, these registers can be unified to avoid repeated changes. what do you think?

sunyuechi commented on erasure_code/riscv64/gf_vect_mul_rvv.S:
```diff
@@ -0,0 +1,114 @@
+.type gf_vect_mul_rvv, @function
+
+/* Function arguments:
+ *   a0: len    - Length of vector in bytes. Must be aligned to 32B.
```
align should refer to address; here it’s merely a multiple of 32

sunyuechi commented on erasure_code/riscv64/gf_vect_mul_rvv.S:
```diff
@@ -0,0 +1,114 @@
+    andi x_tmp, x_len, 0x1F
+    bnez x_tmp, .return_fail
+
+    vsetvli t6, x_len, e8, m1  /* Set vector length to 4 bytes (for mask) */
```
What do the comments "4 bytes" and "for mask" refer to here?  removed?

lvshuo2016 replied:
Removed.

lvshuo2016 commented on erasure_code/riscv64/gf_5vect_mad_rvv.S:
```diff
@@ -0,0 +1,226 @@
+.return_pass:
+    ld s8, 0(sp)
+    ld s9, 8(sp)
+    ld s10, 16(sp)
```
Modified.

sunyuechi commented on erasure_code/riscv64/gf_6vect_dot_prod_rvv.S:
```diff
@@ -0,0 +1,283 @@
+    ld x_dest6, 40(x_dest)
+
+.Llooprvv_vl:
+#    vsetvli s8, x_len, e8, m1
```
unused

lvshuo2016 replied:
Removed.

sunyuechi commented on erasure_code/riscv64/gf_4vect_dot_prod_rvv.S:
```diff
@@ -0,0 +1,227 @@
+
+/* Loop 1: x_len, vector length */
+.Lloop_rvv_vl:
+    vsetvli t0, x_len, e8, m1  /* Set vector length based on x_len */
```
removed?

lvshuo2016 replied:
Removed.

sunyuechi replied:
The same issue exists in other vect_dot_prod files as well. Please remove it from all of them.

lvshuo2016 replied:
Removed all.

sunyuechi commented on erasure_code/riscv64/gf_4vect_dot_prod_rvv.S:
```diff
@@ -0,0 +1,227 @@
+    bge x_pos, x_len, .return_pass
+
+    /* Increment position */
+    #add x_pos, x_pos, t0
```
unused

lvshuo2016 replied:
Removed.

sunyuechi commented on erasure_code/riscv64/gf_4vect_dot_prod_rvv.S:
```diff
@@ -0,0 +1,227 @@
+    ld s2, 16(sp)
+    ld s3, 24(sp)
+    ld s4, 32(sp)
+    ld s8, 40(sp)
```
unused

lvshuo2016 replied:
Is it better to keep the stack on RV as much as possible 16B alignment?

sunyuechi replied:
Yes, you can leave the sp unchanged—just delete the extra ld and sd instructions.

sunyuechi commented on erasure_code/riscv64/gf_6vect_dot_prod_rvv.S:
```diff
@@ -0,0 +1,283 @@
+    ld s8, 64(sp)
+    ld s9, 72(sp)
+    ld s10, 80(sp)
+    ld s11, 88(sp)
```
unused

lvshuo2016 replied:
Is it better to keep the stack on RV as much as possible 16B alignment?

lvshuo2016 replied:
The ld and sd instructions is deleted.

pablodelara commented:
> The self-hosted runner lost communication with the server. Verify the machine is running and has a healthy network connection. Anything in your workflow that terminates the runner process, starves it for CPU/Memory, or blocks its network access can cause this error. It looks the test failed again. The test case did not report an error. In my local machine, It has run successfully many time.

Feel free to push the latest changes and see if the issue is gone on the RISC-V system.
@sunyuechi, are you expecting further review? I appreciate your review very much, but it looks like the comments are usually targeting multiple files, which tells me you are reviewing these in parts. Maybe better to complete the review in one go, so @lvshuo2016 can address it and close this PR soon. Many thanks!

sunyuechi commented:
I have restarted the GitHub Actions runner. Please try again.

> Maybe better to complete the review in one go

@pablodelara 
I'll try to do that. but it's not always easy to catch everything in one pass.
If the review is complete, I’ll make that clear.

sunyuechi commented on erasure_code/riscv64/gf_3vect_dot_prod_rvv.S:
```diff
@@ -0,0 +1,198 @@
+    ret
+
+.return_fail:
+    li a0, 1                # 返回 1 (失败)
```
Remove Chinese comments

sunyuechi commented on erasure_code/riscv64/gf_3vect_dot_prod_rvv.S:
```diff
@@ -0,0 +1,198 @@
+    add x_dest2,x_dest2, s8
+    add x_dest3,x_dest3, s8
+
+    add x_pos, x_pos, s8         # x_pos += 16
```
Is it not necessarily 16 here?

sunyuechi commented on erasure_code/riscv64/gf_vect_mul_rvv.S:
```diff
@@ -0,0 +1,114 @@
+#define v_gft1_hi v8
+
+gf_vect_mul_rvv:
+    /* Check if len is aligned to 32 bytes */
```
~~aligned~~

lvshuo2016 replied:
Removed.

lvshuo2016 commented on erasure_code/riscv64/gf_3vect_mad_rvv.S:
```diff
@@ -0,0 +1,182 @@
+    addi sp, sp, -24
+    sd s8, 0(sp)
+    sd s9, 8(sp)
+    sd s10, 16(sp)
```
One of the main considerations in the code is readability, as other dotprod files s8-s10 all serve the same function. If you insist that s8–s10 can be replaced with t4–t6, I will modify it.

lvshuo2016 commented on erasure_code/riscv64/gf_5vect_mad_rvv.S:
```diff
@@ -0,0 +1,226 @@
+    sd s10, 16(sp)
+
+
+    vsetvli s9, x0, e8, m1
```
add x_pos, x_pos, s9   s9 is used.  s10 is replaced with t6.

lvshuo2016 replied:
Please @pablodelara  take a look.

pablodelara replied:
I am talking from x86 asm experience here, my knowledge in RISC-V is way more limited. Avoid using unnecessarily the stack is generally a good practice, but having said this, the impact in performance and lines of code is very insignificant, from my point of view. I personally don't see a big deal leaving it as it is.

lvshuo2016 commented:
I fixed the case that went wrong last time. Because my development machine and the machine where I submitted the code were not in the same network segment, the development machine could not access the Internet, and their codes were inconsistent last time. I replaced the patch manually. If there are any replacement errors, I will fix it.

lvshuo2016 commented:
> As we discussed earlier, some `s` can be renamed to `t` regs to eliminate stack operations. In the latest commit, `gf_3vect_mad_rvv.S` modified some register names, but the stack operations were not removed (`s8/s9` -> `t4`).
> 
> `gf_3vect_dot_prod_rvv.S` is similar — at least one `s` register can be replaced with `t0`.
> 
> There may be other files as well where temporary variables used only before the loop are reused inside the loop to eliminate stack operations.

In gf_3vect_mad_rvv.S file,there is no stack operations in the latest commit. 

In gf_3vect_dot_prod_rvv.S, s9 has been modified with t0 to reuse temporary variables. There is no similar operation in other dot_prod 4-7 file.


 In addition, do you have any other questions? If so, please let me know and I can fix them as soon as possible.

lvshuo2016 commented on erasure_code/riscv64/gf_4vect_dot_prod_rvv.S:
```diff
@@ -0,0 +1,220 @@
+    sd s9, 48(sp)
+    sd s10, 56(sp)
+
+    vsetvli t0, x_len, e8, m1  /* Set vector length to maximum */
```
It is sufficient for practical storage application, and the storage system use  x_len is at least 4K.  Anyway, I change it to x0

lvshuo2016 commented:
All is updated.

lvshuo2016 commented:
@sunyuechi hi yuechi, do you have any further comments? Thanks!

sunyuechi commented on erasure_code/riscv64/gf_3vect_dot_prod_rvv.S:
```diff
@@ -0,0 +1,188 @@
+    slli t_offset, x_vec, 5
+    ld x_dest1, 0(x_dest)
+    ld x_dest2, 8(x_dest)
+    ld x_dest3, 16(x_dest)
```
This is the last usage of x_dest, so x_dest and x_dest3 can share the same reg to reduce stack operations.
Please check for similar issues.

sunyuechi commented on erasure_code/riscv64/gf_3vect_dot_prod_rvv.S:
```diff
@@ -0,0 +1,188 @@
+    slli t0, x_vec_i, 3
+    add t0,x_src,t0
+    ld x_ptr, 0(t0)
+    add x_ptr,x_ptr,x_pos
```
t0 → x_ptr, then replace one s reg with t0 to reduce stack operations.
Please check for similar issues.

sunyuechi commented on erasure_code/riscv64/gf_3vect_dot_prod_rvv.S:
```diff
@@ -0,0 +1,188 @@
+    vle8.v v_gft3_hi, (x_tbl3)
+    addi x_tbl3, x_tbl3, 16
+
+/* dest 1 */
```
indent

sunyuechi commented:
@pablodelara Sure, I have no problem with that.
